IdentityCore/src/util/NSDictionary+MSIDQueryItems.m

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "NSDictionary+MSIDQueryItems.h"
+
+@implementation NSDictionary (MSIDQueryItems)
+
++ (instancetype)msidDictionaryFromQueryItems:(NSArray<NSURLQueryItem *> *)queryItems
+{
+    NSMutableDictionary *dictionary = [NSMutableDictionary new];
+    for (NSURLQueryItem *item in queryItems) dictionary[item.name] = item.value;
+    
+    return dictionary;
+}
+
+- (NSArray<NSURLQueryItem *> *)msidQueryItems
+{
+    NSMutableArray *items = [NSMutableArray new];
+    [self enumerateKeysAndObjectsUsingBlock:^(id key, id value, BOOL *stop)
+    {
+        if (![[key class] isEqual:NSString.class])
+        {
+            MSID_LOG_WITH_CORR_PII(MSIDLogLevelWarning, nil, @"Failed to create NSURLQueryItem from dictionary: key '%@' is not a string.", key);
+            
+            [items removeAllObjects];
```

Comment:
should we just skip wrong values?

Reply:
I am worried we could end up passing invalid (incompleted) request in such case. On this level we don't know which properties or required and which are optional.

Reply:
Agree with the worry. Perhaps we should follow the current model in the broker request where some properties are specified as mandatory and some not. In this case, request that understands broker properties would need to verify that all required properties are present? Otherwise, I'm a bit worried that we'll fail for non-essential fields.

Reply:
updated